src/Simulation/qdk_sim_rs/src/linalg/expm.rs

```diff
@@ -0,0 +1,212 @@
+    ///
+    /// [`expm`]: Expm::expm
+    type Error;
+    /// Type used to represent $e^{A}$ when calling [`expm`].
+    ///
+    /// [`expm`]: Expm::expm
+    type Output;
+
+    /// Returns the matrix exponential $e^{A}$ of a matrix $A$.
+    fn expm(&self) -> Result<Self::Output, Self::Error>;
+
+    /// Returns the matrix exponential $e^{\alpha A}$ of a matrix $A$,
+    /// and where $\alpha$ is a complex scalar.
+    fn expm_scale(&self, scale: A) -> Result<Self::Output, Self::Error>;
+}
+
+impl<A> Expm<A> for ExplicitEigenvalueDecomposition<A>
+where
+    A: Scalar + Mul<Output = A>,
+    ExplicitEigenvalueDecomposition<A>: EigenvalueDecomposition<A>,
+{
+    type Error = <Self as EigenvalueDecomposition<A>>::Error;
+    type Output = <Self as EigenvalueDecomposition<A>>::MatrixSolution;
+
+    fn expm(&self) -> Result<Self::Output, Self::Error> {
+        self.apply_mtx_fn(|x| x.exp())
+    }
+
+    fn expm_scale(&self, scale: A) -> Result<Self::Output, Self::Error> {
+        self.apply_mtx_fn(|x| (*x * scale).exp())
+    }
+}
+
+#[cfg(feature = "pade")]
+impl<A, S, E> Expm<A> for ArrayBase<S, Ix2>
+where
+    S: Data + RawData<Elem = A>,
+    A: Scalar + ScalarOperand,
+    E: std::convert::From<QdkSimError>,
+    ArrayBase<S, Ix2>: Inv<Error = E> + Identity,
+{
+    type Error = E;
+    type Output = ArrayBase<OwnedRepr<A>, Ix2>;
+
+    fn expm(&self) -> Result<Self::Output, Self::Error> {
+        // TODO: allow generalizing p, q
+        let p = 16;
+        let q = 16;
+
+        let n = pade_n::<A, S, E>(self, p, q)?;
+        let d_inv = pade_d::<A, S, E>(self, p, q)?.inv()?;
+
+        Ok(d_inv.dot(&n))
+    }
+
+    fn expm_scale(&self, scale: A) -> Result<Self::Output, Self::Error> {
+        Ok((self * scale).expm()?)
+    }
+}
+
+// TODO: Allow simpler type constraints.
+// fn pade_d<'a, T: Identity + MatrixPower<Error = E>, E>(mtx: &'a T, p: u32, q: u32) -> Result<T, E>
+// where &'a T: Neg<Output = T> {
+#[cfg(feature = "pade")]
+fn pade_d<A: Scalar + ScalarOperand, S: Data + RawData<Elem = A>, E>(
+    mtx: &ArrayBase<S, Ix2>,
+    p: u32,
+    q: u32,
+) -> Result<ArrayBase<OwnedRepr<A>, Ix2>, E>
+where
+    ArrayBase<S, Ix2>: Inv<Error = E> + Identity,
+    E: std::convert::From<QdkSimError>,
+{
+    // D_pq(A) ≔ Σ_j=0^q [(p + q - j)! p!] / [(p + q)! j! (q - j)!] (-A)^j
+    let mut result = mtx.eye_like().to_owned();
+    for j in 0..q + 1 {
+        let arr: [i64; 0] = [];
+        assert!(arr.len() == 0);
+        let coeff = (approximate_factorial::<_, f64>(p + q - j)
+            * approximate_factorial::<_, f64>(p))
+            / (approximate_factorial::<_, f64>(p + q)
+                * approximate_factorial::<_, f64>(j)
+                * approximate_factorial::<_, f64>(q - j));
+        let coeff = A::from(coeff).unwrap();
+        let mtx_pow = (-mtx.to_owned()).matrix_power(j.into())?;
+        let term = mtx_pow * coeff;
+        result = result + term;
+    }
```

Comment:
I see that `arr` is created, but I fail to see how it is used here in this scope. What shall we lose if we remove lines 120 - 121?

Reply:
Good catch, thanks; I think that was from older attempts that I forgot to remove while refactoring.